panels/updates/cc-updates-panel.c

```diff
@@ -0,0 +1,688 @@
+
+      gtk_label_set_label (GTK_LABEL (self->network_name_label), title);
+    }
+  else
+    {
+      if (connection)
+        gtk_label_set_label (GTK_LABEL (self->network_name_label), _("Connected"));
+      else
+        gtk_label_set_label (GTK_LABEL (self->network_name_label), _("No active connection"));
+    }
+
+  /* Metered status */
+  gtk_widget_set_visible (self->metered_data_label, connection != NULL);
+
+  if (connection)
+    {
+      NMSettingConnection *setting;
+      NMMetered metered;
+
+      setting = nm_connection_get_setting_connection (connection);
+      metered = nm_setting_connection_get_metered (setting);
+
+      if (metered == NM_METERED_YES || metered == NM_METERED_GUESS_YES)
+        gtk_label_set_label (GTK_LABEL (self->metered_data_label), _("Limited data plan connection"));
+      else
+        gtk_label_set_label (GTK_LABEL (self->metered_data_label), _("Unlimited data plan connection"));
+    }
+
+  /* Automatic Updates switch */
+  gtk_widget_set_sensitive (self->automatic_updates_container, connection != NULL);
+
+  ensure_setting_user (self, connection);
+
+  if (connection)
+    {
+      NMSettingUser *setting_user;
+      const gchar *value;
+
+      setting_user = NM_SETTING_USER (nm_connection_get_setting (connection, NM_TYPE_SETTING_USER));
+      value = nm_setting_user_get_data (setting_user, NM_SETTING_ALLOW_DOWNLOADS_WHEN_METERED);
+
+      g_signal_handlers_block_by_func (self->automatic_updates_switch,
+                                       on_automatic_updates_switch_changed_cb,
+                                       self);
+
+      gtk_switch_set_active (GTK_SWITCH (self->automatic_updates_switch), g_strcmp0 (value, "1") == 0);
+
+      g_signal_handlers_unblock_by_func (self->automatic_updates_switch,
+                                         on_automatic_updates_switch_changed_cb,
+                                         self);
+    }
+}
+
+static void
+load_tariff (CcUpdatesPanel *self)
+{
+  g_autoptr(MwtTariffLoader) tariff_loader = NULL;
+  g_autoptr(GInputStream) input_stream = NULL;
+  g_autoptr(GBytes) bytes = NULL;
+  g_autoptr(GFile) tariff_file = NULL;
+  g_autoptr(GError) error = NULL;
+  g_autofree gchar *tariff_path = NULL;
+
+  g_debug ("Loading tariff");
+
+  tariff_path = g_build_filename (g_get_user_config_dir (), SYSTEM_TARIFF_FILENAME, NULL);
+
+  if (!g_file_test (tariff_path, G_FILE_TEST_EXISTS))
+    {
+      g_debug ("Could not open test tariff file");
+      return;
+    }
+
+  if (!g_file_test (tariff_path, G_FILE_TEST_IS_REGULAR))
+    {
+      g_debug ("Test tariff file is not a regular file");
+      return;
+    }
+
+  tariff_file = g_file_new_for_path (tariff_path);
+  input_stream = G_INPUT_STREAM (g_file_read (tariff_file, NULL, &error));
+  if (error)
+    {
+      g_warning ("Error reading test tariff file: %s", error->message);
+      return;
+    }
+
+  bytes = g_input_stream_read_bytes (input_stream, 8192, NULL, &error);
```

Comment:
This should be async. Or, if you don’t want it to be async, just use `g_file_get_contents()` and `g_bytes_new_take()`.

Reply:
Actually, I was thinking. `mogwai-scheduled` needs to be able to read the tariff for each connection in order to know when to allow downloads. So either it needs access to `~/.config/system-tariff` (which is not a good idea from a sandboxing point of view: a system daemon with access to a user’s home directory; it also doesn’t work if the connection is shared between multiple users by NM), or the tariffs need to be stored by `mogwai-scheduled`. I don’t particularly fancy writing multi-user storage code in `mogwai-scheduled`, but then I wondered:

We could just store the tariff on the `NMSettingUser`, in addition to the property saying whether to download when the connection is metered. The `NMSettingUser` code requires that settings values are 8KiB or smaller, and valid UTF-8 (see `nm_setting_user_check_val()`). That should be fine for storing a tariff, if you take the `GVariant` format of the tariff, and convert it to/from UTF-8 using `g_variant_print()` and `g_variant_parse()`.

That should simplify the async handling here too, since you’d no longer be dealing with files; just with `NMSettingUser`. What do you think?

Reply:
One complication here is that, currently, the UI assumes the tariff applies to all NM connections. The approach I’m suggesting applies it only to the currently active connection. Which do we want? Eventually, presumably, the UI will need to allow different tariffs to be configured for different connections; what compromise solution would be best here?